File: backend/app/schemas/workflow.py
from __future__ import annotations
from typing import Any, Literal
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, ConfigDict


class NodeConfig(BaseModel):
    id: str
    type: Literal["input", "tool", "prompt"]
    order: int
    config: dict[str, Any]


class WorkflowCreate(BaseModel):
    name: str
    nodes: list[NodeConfig] = []


class WorkflowUpdate(BaseModel):
    name: str
    nodes: list[NodeConfig]


class WorkflowRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str
    nodes: list[NodeConfig]
    created_at: datetime
    updated_at: datetime


class WorkflowSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str
    updated_at: datetime


class ExecuteRequest(BaseModel):
    input_overrides: dict[str, str] = {}


class StepResult(BaseModel):
    node_id: str
    type: str
    variable: str
    output: str
    duration_ms: int
    error: str | None = None
    tool_name: str | None = None


class ExecuteResponse(BaseModel):
    workflow_id: str
    status: Literal["completed", "failed"]
    steps: list[StepResult]

File: frontend/src/types/workflow.ts
export type NodeType = 'input' | 'tool' | 'prompt';

export interface InputNodeConfig {
  variable_name: string;
  default_value: string;
}

export interface ToolNodeConfig {
  tool_name: string;
  input_variable: string;
  output_variable: string;
}

export interface PromptNodeConfig {
  prompt_template: string;
  output_variable: string;
}

export interface WorkflowNode {
  id: string;
  type: NodeType;
  order: number;
  config: InputNodeConfig | ToolNodeConfig | PromptNodeConfig;
}

export interface WorkflowCreate {
  name: string;
  nodes: WorkflowNode[];
}

export interface WorkflowSummary {
  id: string;
  name: string;
  updated_at: string;
}

export interface WorkflowRead {
  id: string;
  name: string;
  nodes: WorkflowNode[];
  created_at: string;
  updated_at: string;
}

export interface StepResult {
  node_id: string;
  type: string;
  variable: string;
  output: string;
  duration_ms: number;
  error?: string;
  tool_name?: string;
}

export interface ExecuteResponse {
  workflow_id: string;
  status: 'completed' | 'failed';
  steps: StepResult[];
}

export interface ExecuteRequest {
  input_overrides: Record<string, string>;
}
